### PR TITLE
JDK 21 move

### PR DESCRIPTION
**Pull request by andsel**
## Release notes

Adaptations to run Logstash on JDK 21

## What does this PR do?

Adaptations to JDK 21:
- Java 8 support is obsolete and will be removed.
- Thread's `getId` (not final) replaced by final `threadId` https://bugs.openjdk.org/browse/JDK-8017617
- Verify the warnings "this-escape" when a constructor use other method or pass around `this` reference to other methods https://bugs.openjdk.org/browse/JDK-8015831
- URL constructor is deprecated, use `<uri_instance>.toURL()` (since JDK 20)
-  Manages new (since JDK 20) `G1 Concurrent GC` MX Bean, [ref](https://github.com/elastic/logstash/pull/15719#issuecomment-1946367785)

## Why is it important/What is the impact to the user?

As a developer I want that all the blockings to use JDK 21 are removed before effectively switching the bundled JDK to version 21

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~

## Author's Checklist

- [x] run Buildkite JDK 21 matrix for this PR: https://buildkite.com/elastic/logstash-linux-jdk-matrix-pipeline/builds/136

## How to test this PR locally

Install a JDK 21 on your host, example:
```sh
sdk install java 21.0.2-tem
```
and run some testing, like
```sh
./gradlew runIntegrationTests -PrubyIntegrationSpecs="specs/mixed_codec_spec.rb" --console=plain
```

or if you would run all integration tests:
```sh
export BUILD_JAVA_HOME=/<home>/.sdkman/candidates/java/21.0.2-tem && export RUNTIME_JAVA_HOME=/<home>/.sdkman/candidates/java/21.0.2-tem && export LS_JAVA_HOME=/<home>/.sdkman/candidates/java/21.0.2-tem
ci/integration_tests.sh
```
try also a run of Logstash with `LS_JAVA_HOME` set to a JDK 21 path.

## Related issues

- Relates #15342 
- Requires #15707

### CODE REVIEW

File: tools/jvm-options-parser/build.gradle [hunk 1]
```diff
@@ -31,8 +31,8 @@ buildscript {
   }
 }
 
-project.sourceCompatibility = JavaVersion.VERSION_1_8
-project.targetCompatibility = JavaVersion.VERSION_1_8
+project.sourceCompatibility = JavaVersion.VERSION_11
+project.targetCompatibility = JavaVersion.VERSION_11
```
Comment (andsel) [inline tools/jvm-options-parser/build.gradle:35]: **Note for reviewer**

Running Gradle 8.5 with JDK 21 with
```sh
./gradlew :jvm-options-parser:compileJava
```
trigger the error:
```
> Task :jvm-options-parser:compileJava FAILED
warning: [options] source value 8 is obsolete and will be removed in a future release
warning: [options] target value 8 is obsolete and will be removed in a future release
warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
error: warnings found and -Werror specified

FAILURE: Build failed with an exception.
```

Java 8 is obsolete and its support will be dropped

**Comment by andsel** [comment]
As reported in [Oracle blog](https://blogs.oracle.com/javamagazine/post/java-20-gems-jdks), a new `G1 Concurrent GC` JMX Bean is published to track `remark` and `clean` phases times spent.

```
GarbageCollectorMXBean for remark and cleanup pause time in G1. The Garbage-First (G1) garbage collector now has a new GarbageCollectorMXBean called G1 Concurrent GC. This bean reports the duration and occurrence of the remark and cleanup garbage collection pauses.

These pauses occur during a complete concurrent mark cycle, increasing the collection counter by two (one for the remark pause and one for the cleanup pause), similar to the CGC field of the jstat -gcutil command. During these pauses, G1 Concurrent GC also updates the memory pool for the G1 Old Gen MemoryManagerMXBean.
```

File: lib/bootstrap/rspec.rb [hunk 2]
```diff
@@ -15,6 +15,12 @@
 # specific language governing permissions and limitations
 # under the License.
 
+# The following line is needed because from starting from JDK 21 LinkedHashSet
+# has a package private map method that JRuby bound as a usable method.
+# It's mainly related to Gradle opening java.base module and this interfere with
+# JRuby binding of methods. Full description at https://github.com/jruby/jruby/issues/8061#issuecomment-1908807511
+java.util.LinkedHashSet.remove_method(:map) rescue nil
```
[the remark this answers is not in the document]
Comment (andsel) [reply lib/bootstrap/rspec.rb:23]: Fixed by 3b520e2839fde8dfb8bad03fca468635403e3e1b

**Review by jsvd — APPROVED** [review]
Tested locally, works correctly. LGTM

File: lib/bootstrap/rspec.rb [hunk 3]
```diff
@@ -15,6 +15,13 @@
 # specific language governing permissions and limitations
 # under the License.
 
+# The following line is a workaround needed because starting from JDK 21 LinkedHashSet
+# has a package private map method that JRuby bound as a usable method.
+# It's mainly related to Gradle opening java.base module and this interfere with
+# JRuby binding of methods. Full description at https://github.com/jruby/jruby/issues/8061#issuecomment-1908807511
+# Remove when Logstash bundle a JRuby version >= 9.4.6.0.
+java.util.LinkedHashSet.remove_method(:map) rescue nil
```
Comment (jsvd) [inline lib/bootstrap/rspec.rb:23]: Actually now that main and 8.13 are on JRuby 9.4.6.0 we should be able to remove this.

**Review by jsvd — COMMENTED** [review]
We should be able to remove the JRuby < 9.4.6.0 workaround since main/8.13 is on that version.

File: lib/bootstrap/rspec.rb (hunk 3 above)
Comment (andsel) [reply lib/bootstrap/rspec.rb]: @jsvd yep, good spot :-) removed with 1f5f73572d7ef2c140484c6b9e75dbb8886b08ef and it's green.
Ready for a quick review

**Review by jsvd — APPROVED** [review]
LGTM

**Pull request merged**